File: apps/web/src/types/graph.ts
export interface GraphNodeDTO {
  id: string;
  type: string;
  label: string;
  repository_version_id: string;
  metadata: Record<string, any>;
}

export interface GraphEdgeDTO {
  id: string;
  type: string;
  source: string;
  target: string;
  metadata: Record<string, any>;
}

export interface GraphDTO {
  nodes: GraphNodeDTO[];
  edges: GraphEdgeDTO[];
}

export interface GraphOverviewDTO {
  repository_version_id: string;
  total_nodes: number;
  total_edges: number;
  directories: number;
  files: number;
  classes: number;
  functions: number;
  methods: number;
  variables: number;
  parameters: number;
  external_packages: number;
  imports: number;
  calls: number;
  inheritance_relationships: number;
}

export interface NodeRelationshipDTO {
  id?: string;
  type: string;
  connected_node_id: string;
  connected_node_name?: string | null;
  connected_node_type?: string | null;
  direction: 'INCOMING' | 'OUTGOING';
}

export interface NodeDetailDTO {
  id: string;
  type: string;
  name: string | null;
  file_path: string | null;
  qualified_name: string | null;
  line_start: number | null;
  line_end: number | null;
  language: string | null;
  description: string | null;
  source_code: string | null;
  github_url: string | null;
  repository_id?: string | null;
  repository_version_id?: string | null;
  commit_sha?: string | null;
  branch?: string | null;
  status?: string | null;
  created_at?: string | null;
  children_count?: number | null;
  properties: Record<string, any>;
  incoming_relationships: NodeRelationshipDTO[];
  outgoing_relationships: NodeRelationshipDTO[];
}

File: apps/api/src/schemas/graph.py
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional

class GraphNodeDTO(BaseModel):
    id: str
    type: str
    label: str
    repository_version_id: str
    metadata: Dict[str, Any] = Field(default_factory=dict)

class GraphEdgeDTO(BaseModel):
    id: str
    type: str
    source: str
    target: str
    metadata: Dict[str, Any] = Field(default_factory=dict)

class GraphDTO(BaseModel):
    nodes: List[GraphNodeDTO]
    edges: List[GraphEdgeDTO]

class GraphOverviewDTO(BaseModel):
    repository_version_id: str
    total_nodes: int
    total_edges: int
    directories: int
    files: int
    classes: int
    functions: int
    methods: int
    variables: int
    parameters: int
    external_packages: int
    imports: int
    calls: int
    inheritance_relationships: int

class NodeRelationshipDTO(BaseModel):
    id: Optional[str] = None
    type: str
    connected_node_id: str
    connected_node_name: Optional[str] = None
    connected_node_type: Optional[str] = None
    direction: str  # 'INCOMING' | 'OUTGOING'

class NodeDetailDTO(BaseModel):
    id: str
    type: str
    name: Optional[str] = None
    file_path: Optional[str] = None
    qualified_name: Optional[str] = None
    line_start: Optional[int] = None
    line_end: Optional[int] = None
    language: Optional[str] = None
    description: Optional[str] = None
    source_code: Optional[str] = None
    github_url: Optional[str] = None
    repository_id: Optional[str] = None
    repository_version_id: Optional[str] = None
    commit_sha: Optional[str] = None
    branch: Optional[str] = None
    status: Optional[str] = None
    created_at: Optional[str] = None
    children_count: Optional[int] = None
    properties: Dict[str, Any] = Field(default_factory=dict)
    incoming_relationships: List[NodeRelationshipDTO] = Field(default_factory=list)
    outgoing_relationships: List[NodeRelationshipDTO] = Field(default_factory=list)


